src/newcharacter.cpp

```diff
@@ -1503,8 +1531,58 @@ int set_scenario(WINDOW *w, player *u, int &points)
     ctxt.register_action("NEXT_TAB");
     ctxt.register_action("SORT");
     ctxt.register_action("HELP_KEYBINDINGS");
+    ctxt.register_action("FILTER");
+
+    bool recalc_scens = true;
+    int scens_length;
+    std::string filterstring = "";
+    std::vector<const scenario *> sorted_scens;
 
     do {
+        if (recalc_scens) {
+            sorted_scens.clear();
+            for (scenmap::const_iterator iter = scenario::begin(); iter != scenario::end(); ++iter) {
+                if (!filterstring.empty()) {
+                    if (!lcmatch(iter->second.gender_appropriate_name(u->male), filterstring)) {
+                        continue;
+                    }
+                }
+                sorted_scens.push_back(&(iter->second));
+            }
+            scens_length = sorted_scens.size();
+            if (scens_length == 0) {
+                popup(_("Nothing found.")); // another case of black box in tiles
+                filterstring.clear();
```

Comment:
> Looks like this will display a popup (requiring an input to close) and clear the filter whenever you mis-input a character.

Well you can edit the filter line before you hit enter (like in the crafting filter). I'll look what it would take to make the tabs work with empty vectors here.